### PR TITLE
Fix #156, exceptions from FORCED_STOP not rethrown

### LINKED ISSUE #156: Exception thrown by objective function is not rethrown by nlopt
The following test fails for me using `] dev NLopt`. I expected the exception from within `obj` to be rethrown to the called on `NLopt.optimize`
```
using Test

x0 = [0.5,0.7]
obj(x) = (error("test"); dot(x,x))
opt = NLopt.Opt(:LN_SBPLX, length(x0))
opt.lower_bounds = [-1,-1]
opt.upper_bounds = [1,1]
opt.min_objective=obj

# @show NLopt.optimize(opt, x0)
@test_throws ErrorException NLopt.optimize(opt, x0)
```

```
Test Failed at /home/kirill/Sandboxes/SportsModelling/test/nlopt_exceptions.jl:11
  Expression: NLopt.optimize(opt, x0)
    Expected: ErrorException
  No exception thrown
```

Instead optimize returns:
```
NLopt.optimize(opt, x0) = (0.0, [0.5, 0.7], :FORCED_STOP)
```

```
julia> versioninfo()
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: AMD Ryzen 7 3700X 8-Core Processor
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, znver2)

(@v1.5) pkg> status
Status `~/.julia/environments/v1.5/Project.toml`
  [764a87c0] BoundaryValueDiffEq v2.7.0
  [336ed68f] CSV v0.8.2
  [a2cac450] ColorBrewer v0.4.0
  [5ae59095] Colors v0.12.6
  [a93c6f00] DataFrames v0.22.2
  [0c46a032] DifferentialEquations v6.16.0
  [31c24e10] Distributions v0.24.10
  [442a2c76] FastGaussQuadrature v0.4.4
  [53c48c17] FixedPointNumbers v0.8.4
  [28b8d3ca] GR v0.53.0
  [cc18c42c] GaussianMixtures v0.3.4
  [7073ff75] IJulia v1.23.1
  [82e4d734] ImageIO v0.4.1
  [6218d12a] ImageMagick v1.1.6
  [02fcd773] ImageTransformations v0.8.8
  [916415d5] Images v0.23.3
  [a98d9a8b] Interpolations v0.13.1
  [76087f3c] NLopt v0.6.2 `~/.julia/dev/NLopt`
  [9b87118b] PackageCompiler v1.2.3
  [91a5bcdd] Plots v1.10.1
  [1fd47b50] QuadGK v2.4.1
  [295af30f] Revise v3.1.9
  [2913bbd2] StatsBase v0.33.2
  [bd369af6] Tables v1.2.2
```

### PR DESCRIPTION
**Pull request by ikirill**
It's not possible to `rethrow` outside a `catch` block (Julia gives an error). It was only checking for errors when ret was INVALID_ARGS which is wrong.

### CODE REVIEW

**Comment by ikirill** [comment]
This doesn't really fix the issue because there is no backtrace saved with the exception so it's a little useless the way this pr is.

File: src/NLopt.jl [hunk 1]
```diff
@@ -216,7 +216,7 @@ function chk(o::Opt, result::Result)
             e = nlopt_exception
             if e !== nothing && !isa(e, ForcedStop)
                 nlopt_exception = nothing
-                rethrow(e)
+                throw(e)
```
Comment (stevengj) [inline src/NLopt.jl:219]: Too bad we lose the original stack trace; I guess in the future we could also potentially save the `catch_backtrace()` and wrap everything in our own exception type that shows this along with the original exception.

Comment (ikirill) [reply src/NLopt.jl:219]: I think wrapping the exception type might break try-catch blocks around the optimize call that expect specific exception types to be thrown.

I tried this: https://github.com/ikirill/NLopt.jl/commit/c7aed006b0cb5433de41f650277a451453b00e42

**Comment by ikirill** [comment]
Sorry I forgot about this PR. I think it works, I've been using my own local NLopt and it throws exceptions with backtraces correctly.

**Comment by ikirill** [comment]
I also added saving backtraces to be shown with the exception. Obviously I can't rethrow the same exception, you have to save it from the function being called by C, and then throw it again on the Julia side once the C function returns. So maybe it's a little awkward.

**Comment by axsk** [comment]
Just found this after haggling with debugging an error inside my objective without any stacktraces.
I hope it works fine and will get merged, thank you for the work @ikirill !

**Comment by gvnwst** [comment]
@ikirill Any chance you've been working on this? It would be incredible to have this functionality.

**Comment by odow** [comment]
Closing in favor of #194

**Pull request closed without merging**